api/ratelimit/v1alpha1/ratelimit_types.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// RateLimit allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type RateLimit struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitList contains a list of RateLimit resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimit `json:"items"`
+}
+
+// RateLimitSpec defines the desired state of RateLimit
+type RateLimitSpec struct {
+	// Type decides the scope for the RateLimits.
+	Type RateLimitType `json:"type"`
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// Valid RateLimitType values are:
+//
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	RateLimitTypeGlobal RateLimitType = "Global"
```

Comment:
`Local` not in scope?
how would `RateLimitValue` working on global?

Reply:
rm'ed it from the initial design to break up the feature into smaller pieces
`RateLimitValue` will apply to requests on a route across all proxies. Internally we will be configuring ratelimit and filters in envoy as well as configuring the rate limit service so these limit values are achieved

Reply:
what about custom RLS backend?

Reply:
thanks for raising this. this design is focused on application developers who would like to use the rate limiting feature w/o configuring the underlying infra for global ratelimit (EG takes care of this for them / batteries are included). Advanced users should be able to configure their custom rate limit services once https://github.com/envoyproxy/gateway/issues/24 is completed

Reply:
during implement my personal project https://github.com/zirain/limiter, I feel it's meaningless set limit value on Gloabal rate limiting. I'm curious how would you guys translate `RateLimitValue`.

Reply:
so `eg` plan to make `ratelimit`(include a redis for db) as part of components, I used to think that as outside component.

Reply:
since the ratelimit service requires a db backend (we're using Redis), we have two options here
* always ask the user to configure the backend db URL with other info (via the [config API](https://github.com/envoyproxy/gateway/blob/main/api/config/v1alpha1/envoygateway_types.go))
* by default if there no DB configuration specified, EG could create a demo Redis deployment using the infra manager and link it to the Rate Limit server. imho this is line with the [project goals](https://github.com/envoyproxy/gateway/blob/main/GOALS.md#batteries-included).
We need to add enough info out there to highlight that this should not be used in production. I'm leaning towards starting with this option since it doesnt require an API change, so its easier to revert this decision if needed.

Reply:
also would appreciate if we kept implementation details outside this PR, and focused on the user facing API instead

Reply:
> We need to add enough info out there to highlight that this should not be used in production.

what about production, how can I work with an existing well-managed RLS?